### ISSUE TITLE
segfault on shutdown in m_spanningtree

### ISSUE THREAD

**Issue by shaggie76** [inspircd/inspircd#1717]
I upgraded to InspIRCd 3.3.0 this week after using 2.0.x for years and one of the things I've noticed is that when stopping instances I will sometimes see a segfault in my syslog like this:

Oct 11 18:51:37 stealth kernel: [274035.466235] inspircd[27599]: segfault at 88 ip 00007f75764f7af6 sp 00007ffdc0f654d0 error 4 in m_spanningtree.so[7f75764c0000+6f000]
Oct 11 18:51:39 stealth kernel: [274037.499285] inspircd[27565]: segfault at 88 ip 00007f0fdd667af6 sp 00007ffd0a1dc9c0 error 4 in m_spanningtree.so[7f0fdd630000+6f000]
Oct 11 18:51:41 stealth kernel: [274039.532206] inspircd[27531]: segfault at 88 ip 00007f2b8af9baf6 sp 00007fffa6da7820 error 4 in m_spanningtree.so[7f2b8af64000+6f000]
Oct 11 18:51:43 stealth kernel: [274041.565502] inspircd[27497]: segfault at 88 ip 00007f21395cdaf6 sp 00007ffdce62e5e0 error 4 in m_spanningtree.so[7f2139596000+6f000]
Oct 11 18:51:45 stealth kernel: [274043.598956] inspircd[27463]: segfault at 88 ip 00007f0a4262caf6 sp 00007ffecbd0c650 error 4 in m_spanningtree.so[7f0a425f5000+6f000]
Oct 11 18:51:47 stealth kernel: [274045.632714] inspircd[27429]: segfault at 88 ip 00007f697267eaf6 sp 00007ffdc80cf2e0 error 4 in m_spanningtree.so[7f6972647000+6f000]

Our servers are arranged in star-formation: one node that doesn't accept client connections to act as a hub plus 5 extra instances for client connections that all link to the hub. This topology has served us well and we've hosted over 130K concurrent connections on a single machine with this technique for scaling across multiple cores.

The reproduction is pretty consistent for me: start 0 1 2 3 4 5, stop 5 4 3 2 1 0 and each one will segfault on the way down.

InspIRCd version: 3.3.0

Historically we had fiddled with the makefiles so that there was debug information generated -- I've debugged coredumps in our module before. I'll fiddle tonight to see if I can get debug information generated again so I can get a callstack for you.

**Comment by SaberUK** [comment]
To build with debug information delete the 'build' directory, set with INSPIRCD_DEBUG=1 in the environment, and then rebuild.

**Comment by shaggie76** [comment]
Cheers!

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f78f5fb8d89 in std::vector<Events::ModuleEventListener*, std::allocator<Events::ModuleEventListener*> >::begin (this=0x88) at /usr/include/c++/7/bits/stl_vector.h:573
573           { return const_iterator(this->_M_impl._M_start); }
(gdb) bt
#0  0x00007f78f5fb8d89 in std::vector<Events::ModuleEventListener*, std::allocator<Events::ModuleEventListener*> >::begin (this=0x88) at /usr/include/c++/7/bits/stl_vector.h:573
#1  0x00007f78f5fb8bb4 in insp::detail::flat_map_base<Events::ModuleEventListener*, Events::ModuleEventProvider::Comp, Events::ModuleEventListener*, std::less<Events::ModuleEventListener*> >::begin (this=0x88) at include/flat_map.h:107
#2  0x00007f78f5fe03ab in CmdBuilder::FireEvent (this=0x7ffeb1cb5480, target=0x559fe290d9f0, cmd=0x7f78f6008935 "SNONOTICE", taglist=...) at /home/glen/InspIRCd5/src/modules/m_spanningtree/misccommands.cpp:34
#3  0x00007f78f5fcb8aa in CmdBuilder::CmdBuilder (this=0x7ffeb1cb5480, cmd=0x7f78f6008935 "SNONOTICE") at /home/glen/InspIRCd5/src/modules/m_spanningtree/commandbuilder.h:54
#4  0x00007f78f5feb23c in SpanningTreeProtocolInterface::SendSNONotice (this=0x559fe28cfad0, snomask=76 'L', text="Server \002irc0.warframe.com\002 split: Remote host closed connection")
    at /home/glen/InspIRCd5/src/modules/m_spanningtree/protocolinterface.cpp:107
#5  0x0000559fe110aa3c in SnomaskManager::WriteGlobalSno (this=0x559fe288cdd0, letter=76 'L', text="Server \002irc0.warframe.com\002 split: Remote host closed connection") at /home/glen/InspIRCd5/src/snomasks.cpp:49
#6  0x00007f78f5ff3af4 in TreeServer::SQuitChild (this=0x559fe290d9f0, server=0x559fe2921040, reason="Remote host closed connection") at /home/glen/InspIRCd5/src/modules/m_spanningtree/treeserver.cpp:168
#7  0x00007f78f5fe74f0 in TreeServer::SQuit (this=0x559fe2921040, reason="Remote host closed connection") at /home/glen/InspIRCd5/src/modules/m_spanningtree/treeserver.h:115
#8  0x00007f78f5ffc3f3 in TreeSocket::Close (this=0x559fe2901eb0) at /home/glen/InspIRCd5/src/modules/m_spanningtree/treesocket2.cpp:418
#9  0x00007f78f5fff0d3 in SpanningTreeUtilities::cull (this=0x559fe290d880) at /home/glen/InspIRCd5/src/modules/m_spanningtree/utils.cpp:118
#10 0x00007f78f5fd1f05 in ModuleSpanningTree::cull (this=0x559fe28ce6e0) at /home/glen/InspIRCd5/src/modules/m_spanningtree/main.cpp:799
#11 0x0000559fe10c31c4 in CullList::Apply (this=0x559fe288b7b0) at /home/glen/InspIRCd5/src/cull_list.cpp:57
#12 0x0000559fe10fe71f in ModuleManager::UnloadAll (this=0x559fe288c620) at /home/glen/InspIRCd5/src/modules.cpp:433
#13 0x0000559fe10cd24b in InspIRCd::Cleanup (this=0x559fe287af60) at /home/glen/InspIRCd5/src/inspircd.cpp:103
#14 0x0000559fe11057b3 in InspIRCd::Exit (this=0x559fe287af60, status=10) at /home/glen/InspIRCd5/src/server.cpp:49
#15 0x0000559fe110574f in InspIRCd::SignalHandler (this=0x559fe287af60, signal=15) at /home/glen/InspIRCd5/src/server.cpp:40
#16 0x0000559fe10d08a3 in InspIRCd::Run (this=0x559fe287af60) at /home/glen/InspIRCd5/src/inspircd.cpp:673
#17 0x0000559fe10d0b8a in main (argc=2, argv=0x7ffeb1cb5b98) at /home/glen/InspIRCd5/src/inspircd.cpp:695

```

**Comment by SaberUK** [comment]
Thanks, i'll look into this soon.

**Comment by SaberUK** [comment]
Should be fixed in 8c0912f, thanks for reporting it!
